### PR TITLE
DCOS-10982:  Add healthchecks for multi containers

### PR DESCRIPTION
**Pull request by Poltergeist**
This introduces HealthChecks for Multi Containers
healthchecksforpods mov

This does include commits from #1502

### CODE REVIEW

**Comment by nLight** [comment]
Let's merge the section itself first...

**Comment by Poltergeist** [comment]
**mesosphere-ci** retest this please

**Comment by Poltergeist** [comment]
**mesosphere-ci** retest this please

**Comment by orlandohohmeier** [comment]
**mesosphere-ci** retest this please 🙏

**Comment by orlandohohmeier** [comment]
**mesosphere-ci** retest this please 🙇

**Comment by orlandohohmeier** [comment]
✨💜  Awesome

**Comment by nLight** [comment]
What am I doing wrong?
dec-07-2016 19-36-20

**Comment by MatApple** [comment]
After adding health checks and then going to review/run and then back again - I lose all my health checks.

**Comment by orlandohohmeier** [comment]
@MatApple we lost a commit during the rebase @Poltergeist just reintroduced it, should work now.

**Comment by MatApple** [comment]
Also, if I'm not doing multi-container, I get an error in the JSON editor when I haven't selected any of the `Advanced Health Check Settings`. Seems these are mandatory now for single-container services but not for multi-container. Is this correct?

**Comment by orlandohohmeier** [comment]
@MatApple This should be fixed once the respective Marathon RAML issues have been addressed. We've filed issues and they should be fixed by tomorrow.

**Comment by nLight** [comment]
I get a warning:
Failed propType: Invalid prop `family` of value `mini` supplied to `Icon`, expected one of ["product","system","tiny"]. Check the render method of `PodContainerServiceFormSection`.

File: plugins/services/src/js/components/forms/MultiContainerHealthChecksFormSection.js [hunk 1]
```diff
@@ -0,0 +1,173 @@
...
+        </FormGroup>
+      </div>
+    );
+  }
+
+  getEndpoints() {
+    const {data} = this.props;
+
+    return data.portDefinitions.map((port, index) => {
+      return (<option value={index}>{port.name || index}</option>);
```
Comment (nLight) [inline plugins/services/src/js/components/forms/MultiContainerHealthChecksFormSection.js]: #nit redundant `()`

File: plugins/services/src/js/reducers/serviceForm/Containers.js [hunk 2]
```diff
@@ -103,6 +156,22 @@ module.exports = {
...
+    if (path[2] === 'healthChecks') {
+      if (newState[path[1]].healthChecks == null) {
+        newState[path[1]].healthChecks = [];
+      }
+
+      if (this.healthChecks[path[1]] == null) {
+        this.healthChecks[path[1]] = {};
+      }
+
+      newState[path[1]].healthChecks = healthChecks.call(
```
Comment (nLight) [inline plugins/services/src/js/reducers/serviceForm/Containers.js]: Maybe adding a comment that we're calling `healthChecks` reducer here

File: plugins/services/src/js/reducers/serviceForm/Containers.js [hunk 3]
```diff
@@ -183,6 +256,22 @@ module.exports = {
       return newState;
     }
 
+    if (path[2] === 'healthChecks') {
```
Comment (nLight) [inline plugins/services/src/js/reducers/serviceForm/Containers.js]: This block looks familiar to me :notsureif:

File: plugins/services/src/js/reducers/serviceForm/MultiContainerHealthChecks.js [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+
+  if (item.protocol != null) {
+    newItem.protocol = item.protocol;
+    if (item.protocol.toUpperCase() === 'COMMAND' && item.command != null) {
+      newItem.command = {
+        command: item.command
+      };
+    }
+
+    if (item.protocol.toUpperCase().replace('HTTPS', 'HTTP') === 'HTTP') {
```
Comment (nLight) [inline plugins/services/src/js/reducers/serviceForm/MultiContainerHealthChecks.js]: so no https allowed?

File: plugins/services/src/js/reducers/serviceForm/Containers.js [hunk 5]
```diff
@@ -37,10 +38,57 @@ function containersParser(state) {
...
+          }
+        }
+
+        [
+          'gracePeriodSeconds',
+          'intervalSeconds',
+          'timeoutSeconds',
+          'maxConsecutiveFailures'
+        ].forEach((key) => {
+          if (item[key] != null) {
```
Comment (nLight) [inline plugins/services/src/js/reducers/serviceForm/Containers.js]: let's do `filter` first

File: plugins/services/src/js/reducers/serviceForm/Containers.js [hunk 6]
```diff
@@ -37,10 +38,57 @@ function containersParser(state) {
...
+            ], item.command.command, SET));
+          }
+        }
+
+        [
+          'gracePeriodSeconds',
+          'intervalSeconds',
+          'timeoutSeconds',
+          'maxConsecutiveFailures'
+        ].forEach((key) => {
```
Comment (nLight) [inline plugins/services/src/js/reducers/serviceForm/Containers.js]: let's extract a constant here

**Comment by Poltergeist** [comment]
@nLight Your comments are addressed.

**Comment by nLight** [comment]
If you squeeze it a bit form goes jumpy, but that's a #nit 

screen shot 2016-12-08 at 18 32 19

**Comment by nLight** [comment]
You know what, I give you my 🏷 
Merge it at will

**Comment by orlandohohmeier** [comment]
@nLight we have the same visual issue with the single container form. Please open an issue so that we can fix it for all the health checks. :)

**Comment by nLight** [comment]
@orlandohohmeier I'd extend the scope of that issue, we have it everywhere

**Pull request merged**